### PR TITLE
Federated Join Tutorial Notebook

### PR DESCRIPTION
**Pull request by richard-engineering**
- [x] Fix bug with Union Mode FilterVariables (if we consider it a bug)
- [x] Make Intersection no filter_cohort consistent with UnionMode
- [x] Address any documentation feedback

Images from Wikipedia

### CODE REVIEW

**Comment by adrish-rhino** [comment]
It would be good to add definitions of various terms in the beginning - "federated result " "centralized result" "joined result" or point to other documentation describing the same.

**Review by yuvalrhino — CHANGES_REQUESTED** [review]
Super helpful!
I left a few comments, mostly about sanitizing things before making them accessible to users.

File: examples/rhino-sdk/federated_join/Federated Join Tutorial.ipynb [hunk 1]
```diff
@@ -0,0 +1,585 @@
...
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "{'uid': 'cb0d29d8-4a86-4421-b6c9-8692ccffa063', 'import_status': 'Complete'}\n",
+      "{'uid': '9c182ef9-1c4c-4591-935d-5ed8c6c2ec30', 'import_status': 'Complete'}\n"
+     ]
+    }
+   ],
+   "source": [
+    "WORKGROUP_UID = \"e590e0fa-ae37-48b3-b50e-c232536cefab\"  # Replace this with your workgroup\n",
```
[the remark this answers is not in the document]
Comment (atonetti-rhinohealth) [reply examples/rhino-sdk/federated_join/Federated Join Tutorial.ipynb:73]: I usually will do something like this for UIDs we need users to enter "XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX"

File: examples/rhino-sdk/federated_join/Federated Join Tutorial.ipynb [hunk 2]
```diff
@@ -0,0 +1,585 @@
...
+    "    project_uid=project.uid,\n",
+    "    workgroup_uid=WORKGROUP_UID,\n",
+    "    data_schema_uid=data_schema_uid,\n",
+    "    csv_filesystem_location=\"./rhinoOnPrem/rhinoAgent/tests/test_data/FederatedJoinFilterCohort.csv\",\n",
+    "    image_filesystem_location=\"\",\n",
+    "    method=\"filesystem\",\n",
+    "    is_data_deidentified=True,\n",
+    "    file_base_path=\"\",\n",
+    ")\n",
+    "filter_cohort_at_lahey = session.cohort.add_cohort(filter_cohort_input)\n",
```
Comment (atonetti-rhinohealth) [inline examples/rhino-sdk/federated_join/Federated Join Tutorial.ipynb:110]: Remove lahey from the vairable name

File: examples/rhino-sdk/federated_join/Federated Join Tutorial.ipynb [hunk 3]
```diff
@@ -0,0 +1,585 @@
...
+    "dataschema = session.data_schema.create_data_schema(dataschema_input)\n",
+    "data_schema_uid = dataschema.uid\n",
+    "\n",
+    "filter_cohort_input = CohortCreateInput(\n",
+    "    name=\"Blood Test Results\",\n",
+    "    description=\"Identifiers with Blood Type, No SPO2 or Gender\",\n",
+    "    project_uid=project.uid,\n",
+    "    workgroup_uid=WORKGROUP_UID,\n",
+    "    data_schema_uid=data_schema_uid,\n",
+    "    csv_filesystem_location=\"./rhinoOnPrem/rhinoAgent/tests/test_data/FederatedJoinFilterCohort.csv\",\n",
```
[the remark this answers is not in the document]
Comment (atonetti-rhinohealth) [reply examples/rhino-sdk/federated_join/Federated Join Tutorial.ipynb:104]: Agree with Yuval. I would make the string something like "{Path in /rhino_data}/FederatedJoinFilterCohort.csv" that way users still know what the files is but also that its on their client and not the one that's provided in the folder

File: examples/rhino-sdk/federated_join/Federated Join Tutorial.ipynb [hunk 4]
```diff
@@ -0,0 +1,585 @@
...
+   "cell_type": "markdown",
+   "id": "59adf095",
+   "metadata": {},
+   "source": [
+    "# Intersection Mode\n",
+    "\n",
+    "The data we are interested in lives at two different sites.\n",
+    "Lahey has information about the patient, age and blood type\n",
+    "MGH has the gender SpO2 values."
+   ]
```
Comment (atonetti-rhinohealth) [inline examples/rhino-sdk/federated_join/Federated Join Tutorial.ipynb:50]: Could you add the steps you are about to do so we have a bit more meta commentary? (i.e. 1. Creating a project, 2. Creating a scheama, etc.)

File: examples/rhino-sdk/federated_join/Federated Join Tutorial.ipynb [hunk 5]
```diff
@@ -0,0 +1,585 @@
...
+   "execution_count": 95,
+   "id": "ae4fe67d",
+   "metadata": {
+    "pycharm": {
+     "name": "#%%\n"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "from rhino_health.lib.metrics import Count, Mean, StandardDeviation"
```
Comment (atonetti-rhinohealth) [inline examples/rhino-sdk/federated_join/Federated Join Tutorial.ipynb:141]: Markdown cell above: Calculating Mean for ...

File: examples/rhino-sdk/federated_join/Federated Join Tutorial.ipynb [hunk 6]
```diff
@@ -0,0 +1,585 @@
...
+     "text": [
+      "SpO2 values in data cohort for male patents > 35 years old\n",
+      "Federeated Mean: 1.0224691358024691\n",
+      "Centralized Mean: 1.0224691358024691\n",
+      "True\n"
+     ]
+    }
+   ],
+   "source": [
+    "print(\"SpO2 values in data cohort for male patents > 35 years old\")\n",
```
Comment (atonetti-rhinohealth) [inline examples/rhino-sdk/federated_join/Federated Join Tutorial.ipynb:236]: Markdown Heading like Calculating Mean ... and comparing to centralized...

File: examples/rhino-sdk/federated_join/Federated Join Tutorial.ipynb [hunk 7]
```diff
@@ -0,0 +1,585 @@
...
+     "output_type": "stream",
+     "text": [
+      "Works with grouping. SpO2 values for both genders >35 years old\n",
+      "{'m': {'mean': 1.0224691358024691}, 'f': {'mean': 1.02375}}\n",
+      "True\n"
+     ]
+    }
+   ],
+   "source": [
+    "print(\"Works with grouping. SpO2 values for both genders >35 years old\")\n",
```
Comment (atonetti-rhinohealth) [inline examples/rhino-sdk/federated_join/Federated Join Tutorial.ipynb:276]: Markdown Heading, explain why this is different from the above

File: examples/rhino-sdk/federated_join/Federated Join Tutorial.ipynb [hunk 8]
```diff
@@ -0,0 +1,585 @@
...
+     "text": [
+      "Filtering on multiple data columns\n",
+      "{'mean': 1.0492307692307692}\n",
+      "1.0492307692307692\n",
+      "True\n"
+     ]
+    }
+   ],
+   "source": [
+    "print(\"Filtering on multiple data columns\")\n",
```
Comment (atonetti-rhinohealth) [inline examples/rhino-sdk/federated_join/Federated Join Tutorial.ipynb:315]: Markdown heading

File: examples/rhino-sdk/federated_join/Federated Join Tutorial.ipynb [hunk 9]
```diff
@@ -0,0 +1,585 @@
...
+    "\n",
+    "configuration = Mean(\n",
+    "    variable=\"SpO2\",\n",
+    "    join_field=\"UID\",\n",
+    "    data_filters=[\n",
+    "        {\n",
+    "            \"filter_column\": \"Age\",\n",
+    "            \"filter_value\": 35,\n",
+    "            \"filter_type\": \">\",\n",
+    "            # For intersection mode, if unspecified apples to both cohorts. \n",
```
Comment (atonetti-rhinohealth) [inline examples/rhino-sdk/federated_join/Federated Join Tutorial.ipynb:325]: I don't know what this means; could you elaborate more? If what is unspecified? Maybe provide a dueling example if you think it's an important enough call out

File: examples/rhino-sdk/federated_join/Federated Join Tutorial.ipynb [hunk 10]
```diff
@@ -0,0 +1,585 @@
...
+    "print(joined_results.output)\n",
+    "federated_result = joined_results.output['mean']\n",
+    "\n",
+    "a = filter_df[\"Age\"] > 35\n",
+    "b = filter_df[\"BloodType\"] == \"a\"\n",
+    "valid_uids = first_data_df.UID.isin(filter_df[a&b].UID)\n",
+    "c = first_data_df[\"Age\"] > 35\n",
+    "d = first_data_df[\"Gender\"] == \"m\"\n",
+    "centralized_result = np.mean(first_data_df.SpO2[valid_uids&c&d])\n",
+    "print(centralized_result)\n",
```
Comment (atonetti-rhinohealth) [inline examples/rhino-sdk/federated_join/Federated Join Tutorial.ipynb:356]: Labels next to what is being printed

File: examples/rhino-sdk/federated_join/Federated Join Tutorial.ipynb [hunk 11]
```diff
@@ -0,0 +1,585 @@
...
+    "print(centralized_result == federated_result)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "aa2bd257",
+   "metadata": {},
+   "source": [
+    "# Union Mode\n",
```
Comment (atonetti-rhinohealth) [inline examples/rhino-sdk/federated_join/Federated Join Tutorial.ipynb:365]: Give me a dscription of what Union Mode is. If there is supporting documentation either in the SDK or on ZenDesk maybe link there

File: examples/rhino-sdk/federated_join/Federated Join Tutorial.ipynb [hunk 12]
```diff
@@ -0,0 +1,585 @@
...
+    "\n",
+    "print(\"Logged In\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "59adf095",
+   "metadata": {},
+   "source": [
+    "# Intersection Mode\n",
```
Comment (atonetti-rhinohealth) [inline examples/rhino-sdk/federated_join/Federated Join Tutorial.ipynb:45]: Tell me what Intersection Mode is and where I can learn more if applicable (links to ZenDesk/SDK docs)

File: examples/rhino-sdk/federated_join/Federated Join Tutorial.ipynb [hunk 13]
```diff
@@ -0,0 +1,585 @@
...
+    "\n",
+    "The data we are interested in lives at two different sites.\n",
+    "Lahey has information about the patient, age and blood type\n",
```
Comment (atonetti-rhinohealth) [inline examples/rhino-sdk/federated_join/Federated Join Tutorial.ipynb:48]: Remove client names

File: examples/rhino-sdk/federated_join/Federated Join Tutorial.ipynb [hunk 14]
```diff
@@ -0,0 +1,585 @@
...
+    ")\n",
+    "filter_cohort_at_lahey = session.cohort.add_cohort(filter_cohort_input)\n",
+    "\n",
+    "first_data_cohort_input = CohortCreateInput(\n",
+    "    name=\"SpO2 Values on 1/1\",\n",
+    "    description=\"Identifiers with SPO2 and Gender no Blood Type\",\n",
+    "    project_uid=project.uid,\n",
+    "    workgroup_uid=WORKGROUP_UID,\n",
+    "    data_schema_uid=data_schema_uid,\n",
+    "    csv_filesystem_location=\"./rhinoOnPrem/rhinoAgent/tests/test_data/FederatedJoinDataCohort.csv\",\n",
```
Comment (atonetti-rhinohealth) [inline examples/rhino-sdk/federated_join/Federated Join Tutorial.ipynb:118]: Change file path

File: examples/rhino-sdk/federated_join/Federated Join Tutorial.ipynb [hunk 15]
```diff
@@ -0,0 +1,585 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 102,
+   "id": "e6a93438",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "second_data_df = pd.read_csv(\"./FederatedJoinUnionCohort1.csv\")\n",
```
Comment (atonetti-rhinohealth) [inline examples/rhino-sdk/federated_join/Federated Join Tutorial.ipynb:432]: Markdown above, telling me what and why you are doing this

File: examples/rhino-sdk/federated_join/Federated Join Tutorial.ipynb [hunk 16]
```diff
@@ -0,0 +1,585 @@
...
+     "text": [
+      "Union mean using data filters\n",
+      "{'mean': 1.0182239382239384}\n",
+      "1.0182239382239384\n",
+      "True\n"
+     ]
+    }
+   ],
+   "source": [
+    "print(\"Union mean using data filters\")\n",
```
Comment (atonetti-rhinohealth) [inline examples/rhino-sdk/federated_join/Federated Join Tutorial.ipynb:456]: Markdown above

File: examples/rhino-sdk/federated_join/Federated Join Tutorial.ipynb [hunk 17]
```diff
@@ -0,0 +1,585 @@
...
+    "\n",
+    "federated_result = joined_results.output['mean']\n",
+    "\n",
+    "one = first_data_df[(first_data_df.Age > 35)]\n",
+    "two = second_data_df[(second_data_df.Age > 35) & (second_data_df.Gender == \"m\") & (~second_data_df.UID.isin(one.UID))]\n",
+    "three = third_data_df[(third_data_df.Age > 35) & (third_data_df.Gender == \"f\") & (~third_data_df.UID.isin(two.UID)) & (~third_data_df.UID.isin(one.UID))]\n",
+    "\n",
+    "centralized_result = np.mean(pd.concat([one.SpO2, two.SpO2, three.SpO2]))\n",
+    "\n",
+    "print(centralized_result)\n",
```
Comment (atonetti-rhinohealth) [inline examples/rhino-sdk/federated_join/Federated Join Tutorial.ipynb:499]: print labels for context

File: examples/rhino-sdk/federated_join/Federated Join Tutorial.ipynb [hunk 18]
```diff
@@ -0,0 +1,585 @@
...
+     "text": [
+      "Union mean using FilterVariable Shorthand\n",
+      "{'mean': 1.0301869158878503}\n",
+      "1.0301869158878503\n",
+      "True\n"
+     ]
+    }
+   ],
+   "source": [
+    "print(\"Union mean using FilterVariable Shorthand\")\n",
```
Comment (atonetti-rhinohealth) [inline examples/rhino-sdk/federated_join/Federated Join Tutorial.ipynb:526]: markdown above

**Review by atonetti-rhinohealth — CHANGES_REQUESTED** [review]

File: examples/rhino-sdk/federated_join/Federated Join Tutorial.ipynb [hunk 19]
```diff
@@ -0,0 +1,585 @@
...
+   "id": "3b107de9",
+   "metadata": {
+    "pycharm": {
+     "name": "#%%\n"
+    },
+    "scrolled": true
+   },
+   "outputs": [
+    {
+     "name": "stdout",
```
[the remark this answers is not in the document]
Comment (richard-engineering) [reply examples/rhino-sdk/federated_join/Federated Join Tutorial.ipynb:15]: We don't want to include sample outputs? There's nothing secretive afaik. Or do you mean change the numbering

File: examples/rhino-sdk/federated_join/Federated Join Tutorial.ipynb (hunk 1 above)
[the remark this answers is not in the document]
Comment (richard-engineering) [reply examples/rhino-sdk/federated_join/Federated Join Tutorial.ipynb:73]: Replaced with alex's suggestion

File: examples/rhino-sdk/federated_join/Federated Join Tutorial.ipynb (hunk 4 above)
Comment (richard-engineering) [reply examples/rhino-sdk/federated_join/Federated Join Tutorial.ipynb:50]: Split cell

**Comment by richard-engineering** [comment]
@yuvalrhino @atonetti-rhinohealth would be great if you can take a look at the changes I made and approve/make more changes.

**Review by yuvalrhino — APPROVED** [review]

**Pull request merged**